sql/core/src/main/scala/org/apache/spark/sql/execution/columnar/InMemoryTableScanExec.scala

```diff
@@ -142,13 +171,16 @@ case class InMemoryTableScanExec(
       val cachedBatchesToScan =
         if (inMemoryPartitionPruningEnabled) {
           cachedBatchIterator.filter { cachedBatch =>
+            val partitionFilter = newPredicate(
```

Comment:
Why did you choose to nest this more deeply inside of the `filter` rather than leaving it where it was in `mapPartitionsInternal`? By moving it here, we'll wind up calling `newPredicate` once per batch rather than once per partition, thereby harming performance.